packages/babel-plugin-transform-for-of/src/index.ts

```diff
@@ -90,15 +90,14 @@ export default declare((api, options: Options) => {
             return;
           }
           const i = scope.generateUidIdentifier("i");
-          let array: t.Identifier | t.ThisExpression =
-            scope.maybeGenerateMemoised(right, true);
-
-          const inits = [t.variableDeclarator(i, t.numericLiteral(0))];
-          if (array) {
-            inits.push(t.variableDeclarator(array, right));
-          } else {
-            array = right as t.Identifier | t.ThisExpression;
-          }
+          const array: t.Identifier =
+            scope.maybeGenerateMemoised(right, true) ||
+            scope.generateUidIdentifier("arr");
+
+          const inits = [
+            t.variableDeclarator(i, t.numericLiteral(0)),
+            t.variableDeclarator(array, right),
+          ];
```

Comment:
```suggestion
          let array: t.Identifier | t.ThisExpression =
            scope.maybeGenerateMemoised(right, true);
          if (
            !array &&
            t.isIdentifier(right) &&
            path.get("body").scope.hasOwnBinding(right.name)
          ) {
            array = scope.generateUidIdentifier("arr");
          }

          const inits = [t.variableDeclarator(i, t.numericLiteral(0))];
          if (array) {
            inits.push(t.variableDeclarator(array, right));
          } else {
            array = right as t.Identifier | t.ThisExpression;
          }
```

Is this enough, to generate a smaller output in most cases?

Reply:
https://github.com/babel/babel/pull/16011/files#diff-609c7a3c3b056af3d84b8aaad0b67281d02e017161b20292196af421e8120d2d

No, unless we analyze binding usages.

Reply:
`maybeGenerateMemoised` does check if a binding is reassigned.